collector/rapl_linux.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !norapl
+
+package collector
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+)
+
+type raplCollector struct {
+	fs sysfs.FS
+}
+
+func init() {
+	registerCollector("rapl", defaultEnabled, NewRaplCollector)
+}
+
+// NewRaplCollector returns a new Collector exposing RAPL metrics.
+func NewRaplCollector() (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err == nil {
+		collector := raplCollector{
+			fs: fs,
+		}
+		return &collector, nil
+	}
+	return nil, err
+}
+
+// Update implements Collector and exposes RAPL related metrics.
+func (c *raplCollector) Update(ch chan<- prometheus.Metric) error {
+	// Error intentionally unchecked, nil zones are fine when platform
+	// doesn't have powercap files present.
+	zones, _ := sysfs.GetRaplZones(c.fs)
```

Comment:
Not sure about not handling the error here.
What about calling sysfs.GetRaplZones once in NewRaplCollector to check if zones exist and otherwise treat this as an error?

Reply:
@SuperQ wanted me to support cases where the related kernel module can be unloaded and loaded. Not much conclusions can be made about the files being missing, they can appear and vanish with the module, and the API returning nil for zones is pretty much business as usual if the dynamic module loading is to be supported.

As it is, this now survives the files missing quite ok. So I would rather leave the code as it is, and maybe improve the comment there if you feel it is lacking in details?

Reply:
@discordianfish  can you take a look at this again maybe together with @SuperQ? I'd rather not change the logic as it is now robust regarding the kernel module (and them files) being present or not. It now produces the metrics even if the module is not loaded when node_exporter is started, provided that the module gets loaded at some point. And it will survive the module going missing for a while. Producing metrics continues when the module is loaded, always.

Reply:
@SuperQ What do you think?

Reply:
Like @uniemimu said, we don't want to call it once for NewRaplCollector in case the module is unloaded/loaded while the exporter is running.

Usually we handle these kinds of things with a `os.IsNotExist(err)`, but it looks like GetRaplZones has a custom error unlike some of the other procfs functions.

Until this is fixed, I think the better thing to do here is capture the error and silently pass the collector.

```suggestion
	zones, err := sysfs.GetRaplZones(c.fs)
	if err != nil {
	  return nil
	}
```